src/pvcam.cpp

```diff
@@ -0,0 +1,1279 @@
+          pixel_format_to_sample_type.find((PL_IMAGE_FORMATS)pixel_format);
+        sample_type != pixel_format_to_sample_type.end()) {
+        properties->pixel_type = sample_type->second;
+    } else {
+        LOGE("Unsupported pixel format: %d", pixel_format);
+    }
+}
+
+void
+PVCamCamera::maybe_get_input_triggers_(CameraProperties* properties) const
+{
+    properties->input_triggers = { 0 };
+}
+
+void
+PVCamCamera::maybe_get_output_triggers_(CameraProperties* properties) const
+{
+    properties->input_triggers = { 0 };
+}
+
+void
+PVCamCamera::query_exposure_time_capabilities_(
+  CameraPropertyMetadata* meta) const
+{
+    meta->exposure_time_us = { 0 };
+    meta->exposure_time_us.type = PropertyType_FixedPrecision;
+
+    if (!is_param_available(PARAM_EXPOSURE_TIME)) {
+        return;
+    }
+
+    meta->exposure_time_us.writable = 1;
+
+    ulong64 min, max;
+    if (!query_param_min_max(PARAM_EXPOSURE_TIME, &min, &max)) {
+        min = max = 0;
+    }
+
+    auto res = get_exposure_time_resolution_to_microseconds_();
+
+    // convert the resolution to microseconds
+    meta->exposure_time_us.low = (float)min * res;
+    meta->exposure_time_us.high = (float)max * res;
+}
+
+void
+PVCamCamera::query_binning_capabilities(CameraPropertyMetadata* meta) const
+{
+    meta->binning = { 0 };
+    meta->binning.type = PropertyType_FixedPrecision;
+
+    if (!is_param_available(PARAM_BINNING_PAR) ||
+        !is_param_available(PARAM_BINNING_SER)) {
+        meta->binning.low = meta->binning.high = 1.f;
+        return;
+    }
+
+    meta->binning.writable = 1;
+
+    int32 par_min, par_max;
+    if (!query_param_min_max(PARAM_BINNING_PAR, &par_min, &par_max)) {
+        par_min = par_max = 1;
+    }
+
+    int32 ser_min, ser_max;
+    if (!query_param_min_max(PARAM_BINNING_SER, &ser_min, &ser_max)) {
+        ser_min = ser_max = 1;
+    }
+
+    meta->binning.low = (float)(std::max(par_min, ser_min));
+    meta->binning.high = (float)(std::min(par_max, ser_max));
+}
+
+void
+PVCamCamera::query_roi_capabilities(CameraPropertyMetadata* meta) const
+{
+    meta->offset = { 0 };
+    meta->offset.x.type = PropertyType_FixedPrecision;
+    meta->offset.y.type = PropertyType_FixedPrecision;
+    meta->shape.x.type = PropertyType_FixedPrecision;
+    meta->shape.y.type = PropertyType_FixedPrecision;
+
+    if (!is_param_available(PARAM_ROI)) {
+        return;
+    }
+
+    rgn_type roi_min, roi_max;
+    if (!query_param_min_max(PARAM_ROI, &roi_min, &roi_max)) {
+        return;
+    }
+
+    meta->offset.x = {
+        .writable = 1,
+        .low = (float)roi_min.s1,
+        .high = (float)roi_max.s1,
+        .type = PropertyType_FixedPrecision,
+    };
+
+    meta->offset.y = {
+        .writable = 1,
+        .low = (float)roi_min.p1,
+        .high = (float)roi_max.p1,
+        .type = PropertyType_FixedPrecision,
+    };
+
+    meta->shape.x = {
+        .writable = 1,
+        .low = (float)(roi_min.s2 - roi_max.s1 + 1),
```

Comment:
Is it guaranteed that `roi_min.s2 >= roi_max.s1`?

I assume `roi_min.s2` is the upper bound on the left x-coordinate of the region and `roi_max.s1` is the lower bound on the right x-coordinate of the region, so I could imagine cases where that condition is not guaranteed (i.e. because i can define small regions that start on the right of the sensor, and small regions that end on the left of the sensor).

In that case, I think this calculation in general could produce surprising values (i.e. a large intermediate unsigned result). Though maybe some/most compilers will handle this more gracefully than I give them credit for?

Reply:
My understanding was that `roi_min.s2` is the lower bound on the right x-coordinate of the region and `roi_max.s1` is the upper bound on the left x-coordinate of the region. One thing I missed in the SDK docs about `ATTR_MIN` and `ATTR_MAX`, though:

> The value is only valid for the following data types: TYPE_INT8, TYPE_UNS8, TYPE_INT16, TYPE_UNS16, TYPE_INT32, TYPE_UNS32, TYPE_INT64, TYPE_UNS64, TYPE_FLT64 and TYPE_BOOLEAN.

I'll have to query other parameters. Thanks for catching it.